---
title: "MOSAIC Organises Segments of Ancestry In Chromosomes"
author: "Michael Salter-Townshend"
date: "`r Sys.Date()`"
output: rmarkdown::pdf_document
fig_caption: yes
urlcolor: blue
header-includes: 
- \usepackage{float}
vignette: >
  %\VignetteIndexEntry{MOSAIC}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.pos = "H")
```

# Introduction
MOSAIC is a tool for modelling multiway admixture using dense genotype data. 
Given a set of potentially admixed haplotypes (targets) and multiple labelled sets of potentially related haplotypes (panels), 
MOSAIC will infer the most recent admixture events occurring in the targets in terms of the panels.

See Salter-Townshend and Myers, 2019 <https://doi.org/10.1534/genetics.119.302139>
for details on the MOSAIC method. 
For results on a large set of human populations see <https://maths.ucd.ie/~mst/MOSAIC/HGDP_browser/>. 

It is not necessary that any of the panels are good surrogates for the unseen mixing groups as MOSAIC will infer parameters controlling:

1. Recombination rates before and after admixture ($\rho$ and $\Pi$ respectively). 

1. The stochastic relationship between panels and ancestral groups ($\mu$). 
   
1. Timings (from coancestry curves) and ancestry proportions ($\alpha$) of the admixture events.

1. Mutation / error rates for the haplotypes ($\theta$).

Phasing improvements in light of the admixture model are performed and local ancestry along the genome is estimated.


## Inputs / Data
The `example_data` folder packaged with MOSAIC contains example human data for chromosomes 18 to 22.

As inputs MOSAIC requires:

- Phased haplotypes for reference panels and target (admixed) individuals (MOSAIC attempts to detect and correct phasing errors in targets): 
these should be named `pop.genofile.chr` where pop is the population (panel) name and chr is the chromosome index.  
All entries should be $0,1,?$ indicating ref, alt, or missing entries respectively. The rows are \#snps the columns are and \#haps. 
**Note that there should be no spaces in these files.**

- A population names file: `sample.names` format unimportant apart from first column should have all the population names.

- SNP files: `snpfile.chr` with \#snps rows and 6 columns comprising rsID, chr, distance, position, allele ?, allele ?. 

- Recombination rates files: `rates.chr` 3 rows of \#sites, position, cumulative recombination rate (in centiMorgans). 

Examination of the files in the `example_data` folder should make the format of each of the above clear. 


## Outputs / Results
1. A folder called `MOSAIC_RESULTS` is required to hold log-files (foo.out) and results (foo.RData).  

1. A folder called `MOSAIC_PLOTS` is required to hold the plots created by default by a MOSAIC run.

1. A folder called `FREQS` is required to hold the frequencies used to compute $F_{st}$ statistics if required.


## Parameters Inferred
There are 4 sets of parameters inferred via EM:

1. $\Pi$: prob. of switching between latent ancestries, including switch to same anc; $A\times A$.
1. $\rho$: prob. of switching haps within each ancestry.
1. $\mu$: copying matrix; $\mu_{ia}$ is  the probability of a donor from group i given ancestry a; $K\times A$ where K is \#donorpops. 
1. $\theta$: error / mutation; vector length A; prob. of a difference b/w copied and copying haps at a locus.


# Simple Simulation Study

A quick and simple simulation that takes about 5 minutes to run can be created using:
```{bash, eval=F, echo=T}
Rscript mosaic.R simulated example_data/ -c 18:22 -n 2 -p "English Mandenka"
```
or equivalently in an interactive *R* session:
```{r, eval=FALSE}
require(MOSAIC)
set.seed(123)
mosaic.result=run_mosaic("simulated","example_data/",chrnos=18:22,A=2,NUMI=3,
			 pops=c("English","Mandenka"),gens=30)
```

Once this has completed, MOSAIC will have done:

* Simulated 2-way admixture using English and Mandenkan chromosomes $30$ generations ago.

* Read in these simulated chromosomes, along with all available reference panels in the `example_data` folder.

* Inferred the model parameters $(\mu, \theta, \Pi, \rho)$ via EM.

* Corrected phasing errors that scramble local ancestry. 

* Estimated 2-way 2-way local ancestry along each chromosome for each admixed individual. 

* Estimated $F_{st}$ between each ancestral group and each ancestral group and each reference panel. 

* Saved key plots to `MOSAIC_PLOTS` as PDFs.

* Saved key results to `MOSAIC_RESULTS`.


## Loading Results
The results can be loaded in an *R* session using:
```{r}
load("MOSAIC_RESULTS/simulated_2way_1-3_18-22_148_60_0.99_100.RData") # model parameters, etc
load("MOSAIC_RESULTS/localanc_simulated_2way_1-3_18-22_148_60_0.99_100.RData") # local ancestry 
```

If MOSAIC has been run within *R* using the ``run_mosaic()`` command above then you can alternatively use
```{r, eval=F}
attach(mosaic.result)
```
to attach the results for further use.

## Plots
After loading (or attaching) the results in *R*,  
^[When the seed is set in \texttt{R} using set.seed(123) then identical plots should be obtained.]
each of the plots can be created within *R* by running:
```{r, eval=F}
plot_all_mosaic(pathout="MOSAIC_PLOTS/",target)
```
to output default plots to the folder `MOSAIC_PLOTS/`. 
Note that this is already run automatically by default in ``run_mosaic()``

### Or you can generate each plot individually:
```{r, fig.height=3}
ord.Mu=plot_Mu(Mu,alpha,NL)
```
Inferred Copying Matrix $\mu$. One ancestry (that generated from English segments) mostly copies haplotypes in the French panel and the other (Mandenkan) mostly copies haplotypes from the Yoruban panel, as expected.
\

```{r, fig.height=3}
ord.Fst=plot_Fst(all_Fst$panels,ord=T)
```
$1-F_{st}$ estimates between each ancestral group and each donor panel; closer population pairs are larger.
\
 
This can be flipped to show the actual $F_{st}$ values by including the `reverse=FALSE` argument:
```{r,fig.height=3}
tmp=plot_Fst(all_Fst$panels,reverse=FALSE)
```
Raw unordered $F_{st}$ values; closer population pairs are smaller.
In this case we've also not reordered the panels in order of closeness under $F_{st}$. 
\
 
Plot co-ancestry curves used to infer event timings:
```{r, fig.height=2.5}
fitcc=plot_coanccurves(acoancs,dr)
```
Inferred Coancestry Curves. `dr` is the gap in genetic distance between successive gridpoints. These coancestry curves are somewhat rough at longer distances as we have only used short chromosomes in the analysis. The black lines are empirical coancestry curves across all target individuals, the light grey are per individual, and the green is the fitted single-event coancestry curve.
\
 
Look at the $2^{nd}$ individuals first chromosome:
```{r}
ch=1
ind=2
dipplot(ch,ind,g.loc[[ch]],ind,localanc,xlab=paste("Position on Chromosome",chrnos[ch]),ylab="estimated")
mp<-axTicks(1,round(axp=c(min(g.loc[[ch]]),max(g.loc[[ch]]),5)))
axis(1,at=mp,labels=signif(mp,3))
```
Local ancestry estimates. The second and third line add details to the axes, etc. 
\

